tests/multiple_values.rs

```diff
@@ -1250,3 +1250,72 @@ fn issue_2229() {
     assert!(m.is_err()); // This panics, because `m.is_err() == false`.
     assert_eq!(m.unwrap_err().kind, ErrorKind::WrongNumberOfValues);
 }
+
+#[test]
+fn value_names_building_num_vals() {
+    let m = App::new("test")
+        .arg(
+            Arg::new("pos")
+                .long("pos")
+                .value_names(&["who", "what", "why"]),
+        )
+        .try_get_matches_from(vec!["myprog", "--pos", "val1", "val2", "val3"]);
+
+    assert!(m.is_ok(), "{:?}", m.unwrap_err().kind);
+    let m = m.unwrap();
+
+    assert_eq!(
+        m.values_of("pos").unwrap().collect::<Vec<_>>(),
+        ["val1", "val2", "val3"]
+    );
+}
+
+#[test]
+fn value_names_building_num_vals_from_usage() {
+    let m = App::new("test")
+        .arg(Arg::from("--pos <who> <what> <why>"))
+        .try_get_matches_from(vec!["myprog", "--pos", "val1", "val2", "val3"]);
+
+    assert!(m.is_ok(), "{:?}", m.unwrap_err().kind);
+    let m = m.unwrap();
+
+    assert_eq!(
+        m.values_of("pos").unwrap().collect::<Vec<_>>(),
+        ["val1", "val2", "val3"]
+    );
+}
+
+#[test]
+fn value_names_building_num_vals_for_positional() {
+    let m = App::new("test")
+        .arg(Arg::new("pos").value_names(&["who", "what", "why"]))
+        .try_get_matches_from(vec!["myprog", "val1", "val2", "val3"]);
+
+    assert!(m.is_ok(), "{:?}", m.unwrap_err().kind);
+    let m = m.unwrap();
+
+    assert_eq!(
+        m.values_of("pos").unwrap().collect::<Vec<_>>(),
+        ["val1", "val2", "val3"]
+    );
+}
+
+#[test]
+fn number_of_values_preferred_over_value_names() {
+    let m = App::new("test")
+        .arg(
+            Arg::new("pos")
+                .long("pos")
+                .number_of_values(4)
+                .value_names(&["who", "what", "why"]),
+        )
+        .try_get_matches_from(vec!["myprog", "--pos", "val1", "val2", "val3", "val4"]);
+
+    assert!(m.is_ok(), "{:?}", m.unwrap_err().kind);
```

Comment:
Should this be an error case?

I've seen 1 `value_name` be repeated for all `num_vals` and I've seen N `value_names` for N `num_vals`.  I'm wondering what the use case is for *more* `value_names` or what we usage when there are `1 < value_names < num_vals`

Reply:
`number_of_values` should be given preference which is why this is not an error case. `value_names` can diverge away from `number_of_values`, but that is a matter for us to decide on how the help output should be rendered. We currently handle some cases but not all of them.

Reply:
I think the most likely case for a mismatch, besides 1:M is the user missed one and that a debug assert would be the best course of action.  Probably not worth deciding in this.  I'll go create an issue.